### ISSUE TITLE
Is there anything like GIL?

### ISSUE THREAD

**Issue by yanchenghust** [go-python/gpython#64]
As we know, CPython implemention has a GIL. 
How about gpython?

**Comment by ncw** [comment]
gpython does not contain a GIL.

However it does not, as yet allow any sort of multiprocessing.

It is my plan to make a threading library which under the hood creates go routines.  At this point it will need some sort of locking...

**Comment by sbinet** [comment]
by `threading` library, you mean exposing a `gpython` module to the Python interpreter, right?

the [skylark](https://godoc.org/github.com/google/skylark) Python-like VM has actually the concept of a `Thread` state that is passed to each API call...

- https://godoc.org/github.com/google/skylark#Thread

WDYT about this?

**Comment by ncw** [comment]
> by `threading` library, you mean exposing a `gpython` module to the Python interpreter, right?

Yes.  My plan was to make a low level module exposing some go features like goroutines, channels, locks etc, and use that low level module to build up something which was API compatible with the `threading` module.

> the [skylark](https://godoc.org/github.com/google/skylark) Python-like VM has actually the concept of a `Thread` state that is passed to each API call...

That sounds like a job for a context passed everywhere.  We've already identified somewhere else that would be very useful!  Go having no thread local storage makes it very difficult to translate some bits of python.

> * https://godoc.org/github.com/google/skylark#Thread
> 
> WDYT about this?

I think we'd almost certainly need something very similar! I'd propose to attach it to a context though.

**Comment by sbinet** [comment]
> I think we'd almost certainly need something very similar! I'd propose to attach it to a context though. 

SGTM. (we should probably make sure extracting informations and/or restoring state from that `context.Context` isn't showing abysmal performances, though.)

**Comment by ncw** [comment]
> SGTM. (we should probably make sure extracting informations and/or restoring state from that `context.Context` isn't showing abysmal performances, though.)

Hmm, yes! AFAICS context is implemented as linked list, so the lookup isn't very efficient.

**Comment by drew-512** [comment]
I suggest to close this issue with the push of `v0.1.0` (multi-context support)

**Comment by drew-512** [comment]
@yanchenghust v0.1.0 introduces `py.Context` which is the best-case scenario for concurrent interpreter execution while also aligned with Go programming patterns.  For example, for each goroutine that needs to service a py script, call:
```
go func() {
    ctx := py.NewContext(opts)
    ...
    py.Run*(ctx, ...)
    ...
    ctx.Close()
}()
```

**Comment by sbinet** [comment]
fixed w/ #158
